include/commissioner/error.hpp

```diff
@@ -42,36 +43,209 @@ namespace ot {
-    kNotFound,        ///< Things not found.
-    kNotImplemented,  ///< Feature not implemented.
-    kOutOfMemory,     ///< Running out of memory.
-    kReject,          ///< Request is rejected.
-    kSecurity,        ///< Security related errors.
-    kTimeout,         ///< Request timeouted.
-    kTransportBusy,   ///< Transport level is now busy.
-    kTransportFailed, ///< Transport level failure.
+    /**
+     * Not an error; returned on success.
+     */
+    kNone = 0,
+
+    /**
+     * The operation was cancelled (typically by the caller).
+     */
+    kCancelled = 1,
+
+    /**
+     * Client specified invalid arguments that are problematic
+     * regardless of the state of the system (e.g., a malformed file name).
+     */
+    kInvalidArgs = 2,
+
+    /**
+     * Invalid CLI command.
+     */
+    kInvalidCommand = 3,
+
+    /**
+     * Timeout before operation could complete.  For operations
+     * that change the state of the system, this error may be returned
+     * even if the operation has completed successfully. For example, a
+     * successful response from a server could have been delayed long
+     * enough for the deadline to expire.
+     */
+    kTimeout = 4,
+
+    /**
+     * Some requested entity (e.g., TLV) was not found.
+     * For privacy reasons, this code *may* be returned when the client
+     * does not have the access right to the entity.
+     */
+    kNotFound = 5,
+
+    /**
+     * Security failures, such as signature validation, message signing
+     * and (D)TLS handshake failure.
+     */
+    kSecurity = 6,
+
+    /**
+     * Operation is not implemented or not supported/enabled in this service.
+     */
+    kUnimplemented = 7,
+
+    /**
+     * Message, TLV or encoded data is in bad format.
+     */
+    kBadFormat = 8,
+
+    /**
+     * The commissioner is busy that current request cannot be processed.
+     * It is mostly that the commissioner is receiving duplicate requests
+     * before finishing the previous one.
+     */
+    kBusy = 9,
+
+    /**
+     * Running out of memory.
+     */
+    kOutOfMemory = 10,
+
+    /**
+     * Read/write to file/socket failed.
+     */
+    kIOError = 11,
+
+    /**
+     * The file/socket is busy and read/write to file/socket will be blocked.
+     */
+    kIOBusy = 12,
+
+    /**
+     * Some entity that we attempted to create (e.g., CoAP resource)
+     * already exists.
+     */
+    kAlreadyExists = 13,
+
+    /**
+     * The operation, transaction or message exchange was aborted.
+     */
+    kAborted = 14,
+
+    /**
+     * The commissioner is not in a valid state that the operation can be processed.
+     */
+
+    kInvalidState = 15,
+
+    /**
+     * The operational has been rejected. For example, petition could be rejected
+     * because of existing active commissioner.
+     */
+    kRejected = 16,
+
+    /**
+     * The error is out of the address space of OT Commissioner.
```

Comment:
What does this mean?

Reply:
For  example, there are mbedtls errors that cannot fit into (convert to) OT Commissioner errors. In that case, this error code will be used and the error message will include the details of the mbedtls error.